src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImpl.java

```diff
@@ -104,8 +141,40 @@ public String getIdentifier() {
         return execute(request, typeOfT, typeofU, null);
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public <T, U> GraphqlResponse<T, U> execute(GraphqlRequest request, Type typeOfT, Type typeofU, RequestOptions options) {
+        Cache<CacheKey, GraphqlResponse<Object, Object>> cache = toActiveCache(request, options);
+        if (cache != null) {
+            CacheKey key = new CacheKey(request, options);
+            try {
+                return (GraphqlResponse<T, U>) cache.get(key, () -> executeImpl(request, typeOfT, typeofU, options));
+            } catch (ExecutionException e) {
+                return null;
+            }
+        }
+        return executeImpl(request, typeOfT, typeofU, options);
+    }
+
+    private Cache<CacheKey, GraphqlResponse<Object, Object>> toActiveCache(GraphqlRequest request, RequestOptions options) {
+        if (caches == null || request.getQuery().trim().startsWith("mutation")) {
```

Comment:
Is `request.getQuery().trim().startsWith("mutation")` a save test? There might be comments in the query (if custom queries are used, not our models).

Reply:
Except if there are comments, yes it's fine. Not sure if we really want to make sure this is also fine if there are comments ... the only 100%-safe option would be to parse the query with a GraphQL parser to ne 100% sure that it's a mutation, but we probably don't want to add a dependency to `graphql-java` or another similar library. Or do you maybe have another idea?

Reply:
I think that case would be highly unlikely, since you cannot have a mutation and a query in the same GraphQL request. Otherwise we could add some regex magic.

Reply:
Yeah, I'm fine with the current approach or regex check. Both approaches are ok.